### DISCUSSION TITLE
CameraControls at WebWorker (ThreeJS & Offscreen Canvas)

### DISCUSSION THREAD

**Discussion by QuisMagni** [yomotsu/camera-controls#479] [Q&A] [answered]
Hi,

what would be your suggestion to get the CameraControls working on a webworker? Using CameraControls on MainThread and sending camera transform data to the worker or sending input events?

Are there plans to add better  out-of-the-box support for webworker in the future?

**Comment by yomotsu** [comment]
Using CameraControls with WebWorker is not intended currently because I cannot imagine such a use case.
What is your usecase?

It would be welcome if someone could send a PR for that anyway.

**Reply by QuisMagni** [reply]
If you want a better performance in your app it might be a good idea to run ThreeJS in a seperate thread (a WebWorker) using an Offscreen Canvas. This becomes necessary if you try to run an expensive game engine like this: https://www.youtube.com/watch?v=IgLx78LMic4 (we are developing this game and it uses ThreeJS and your Camera-Library).

Inside a WebWorker there is no access to DOM-Elements or Events so you have to pass this information by yourself from the main thread to the worker.

Here is a example for running ThreeJS inside a WebWorker: https://threejs.org/examples/webgl_worker_offscreencanvas.html

**Reply by yomotsu** [reply, marked as answer]
Thanks for the details.
It seems worth supporting WebWorker.

Although camera-controls requires a DOM element, I think It's possible with some internal changes, according to 
https://github.com/puxiao/using-orbitcontrols-in-worker/tree/main

1. Prepare a virtual dom element pure JS object that extends an EventDispatcher class.
2. Pass the virtual element to `new CameraControl( camera, virtualElement )`
3. Then sync the virtual element state with a real dom element with worker's postMessage. Also, the virtual element emits virtual pointer events via the real dom.
4. camera-controls just use the pure JS object as the Dom element internally.

**Reply by QuisMagni** [reply]
Awesome - thank you!

**Comment by yomotsu** [comment]
@butzemannbiber 
I added a worker example.
https://yomotsu.github.io/camera-controls/examples/worker.html
I hope this will help your development.

Btw, your project looks nice. When the project is released, please let me know!
I'm looking for nice some use cases of CameraControls.
